VectorImpl generic math tests

moxcodes opened the pull request:
## Proposed changes

Fourth intermediate pull request from #935. 
This one adds a generic math testing utility, and supporting functions. The supporting functions include generalizations to the `get_size` and `get_element` utilities in `Utilities/ContainerHelpers` and an approximation check utility for confirming element wise functions.

The math tests have been reworked so that they rely on fewer additional utilities, so the supplementary utilities in `VectorImplTestHelper.hpp` have been eliminated, along with their tests in `Test_VectorImplTestHelper.hpp`

### Types of changes:

- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

(sorry for reviewers who spent time on the PRs which produced the now eliminated utilities.)

moxcodes committed:
Formatting and IWYU fixes

moxcodes committed:
Add make_with_random_values version

- new version is a convenience overload which takes distribution by value

moxcodes committed:
Add type trait is_complex_of_fundamental<T>

- this helper type trait will cut down on SFINAE and pattern-matching intricacy

moxcodes committed:
Include Math.hpp in DereferenceWrapper.hpp

- this ensures that the extra `double` versions of math functions can't be
  accidentally omitted from the dereference wrapping whenever the `DataVector`
  versions are used.

moxcodes committed:
Remove now unnecessary tuple utilities

moxcodes committed:
Remove now unnecessary `check_vectors`

- new design for the generic math tests no longer require the `check_vectors`
  utility

moxcodes committed:
Generalize get_size and get_element

- allow for a user-defined size or element function to support more elaborate
      (e.g. multidimensional) types
- eliminate SFINAE in favor of a pattern-matching struct

moxcodes committed:
Add element-wise function checking utility

moxcodes committed:
Add math testing function utilities

moxcodes committed:
Rework DataVector math tests with new framework

nilsdeppe commented on src/Utilities/TypeTraits.hpp:
```diff
@@ -1393,6 +1394,26 @@ template <typename T>
+/// is a `std::true_type` if so, and otherwise is a `std::false_type`
+///
+/// \snippet Test_TypeTraits.cpp is_complex_of_fundamental
+template <typename T, typename = cpp17::bool_constant<true>>
```
Have you tried just `bool` instead of `bool_constant`? In general that should work and we should try to do that whenever it doesn't result in hitting the compiler bug. A lot of code already does that.

moxcodes replied:
This one needs to be `bool_constant` if the test is to be in the complex specialization, which was needed to avoid SFINAE.

nilsdeppe replied:
:+1:

nilsdeppe commented on src/Utilities/TypeTraits.hpp:
```diff
@@ -1393,6 +1394,26 @@ template <typename T>
+// type
+template <typename T>
+struct is_complex_of_fundamental<
+    std::complex<T>, cpp17::bool_constant<cpp17::is_fundamental_v<T>>>
```
You can move the `is_fundamental_v` into the default definition, then change the `false_type` and `true_type` locations. That's a more common pattern because it results in less code :)

moxcodes replied:
I can't put the `is_fundamental_v` into the default without causing SFINAE. Note that the test needs access to the `value_type` of the complex.

nilsdeppe replied:
Derp, yep...

nilsdeppe commented on src/Utilities/ContainerHelpers.hpp:
```diff
@@ -3,43 +3,138 @@
-
-/// \cond
-template <class T, Requires<std::is_fundamental<T>::value> = nullptr>
-SPECTRE_ALWAYS_INLINE size_t get_size(const T& /*t*/) noexcept {
```
Wouldn't a simpler impletantation be adding a function:
```cpp
template <typename T, typename SizeFunction>
SPECTRE_ALWAYS_INLINE decltype(auto) get_size(const T& t, SizeFunction size) noexcept {
  return size(t);
}
```
? This way in the case of not specifying a extra size function you get the right behavior and if you do then the `Size` function is invoked. I guess you might still want the special case of passing a size function and `T` being a fundamental. What's not clear to me is what behavior you want if you pass a `SizeFunction` and the object has a `.size()` member. So you won't ever really be able to be truly generic, I think the best you can do is return 1 for fundamentals and the `SizeFunction`/`.size()` in the general case.

moxcodes replied:
For the implementation details in the elementwise approximation, I want a default `Size` and `At` callable anyway, so unless I misunderstand, adding the overload would only add code.

nilsdeppe replied:
I'm saying if you add an overload the default doesn't need to be there because it will fall back to the other case (which is your current default) anyway :)  The default behavior is what currently happens, and if you add an argument that doesn't have a default then the compiler can disambiguate between the two and you get your normal default behavior

moxcodes replied:
I understand the process you're suggesting, but currently the default `At` and `Size` structs are also used in the element-wise test, as defaults in much the same way a possibly user-supplied `Approx` is taken.
So, without these default structs, I would also need to make duplicate default overloads in the element-wise test utility as well, which I suspect is a net code gain as compared to the current. 

Ultimately, I would like these default `At` and `Size` structs somewhere: either here or in the `TestHelpers`, and it seems preferable to put them here and use them as defaults.

nilsdeppe replied:
After looking through the code more I think the current implementation is more difficult to understand than it needs to be and I'm really not a big fan of having global scope structs with the names `At` and `Size` floating around since the name is not overly descriptive in a general context. If you would like them to be the defaults in the tests I'm fine with sticking them in a `detail` namespace there so you don't need the various permutations of different defaults.

moxcodes replied:
I'm fine with that design, though it is notable that doing so does change functionality. The current will always return 1 for a fundamental size and itself for a fundamental at. The suggested change insists that a user-defined function must also handle the fundamental case itself as well. 

I don't mind this much, but it means that the result will be almost identical to the original implementation, with the full at and size function definitions in the `TestHelpers` detail namespace.

Edit: to be clear, I'm struggling to see a case where the function in the original comment would be used. If the calling code has access to the `SizeFunction` that it passes to `get_size`, and `get_size` just always forwards to the size function, why would it not call the size function directly?

nilsdeppe replied:
Fair enough. How about this:
```cpp
/// \ingroup UtilitiesGroup
/// \brief Callable struct which retrieves the `t.size()` for operand `t`. This
/// will cause a compiler error if no such function exists.
struct GetContainerSize {
  template <typename T>
  SPECTRE_ALWAYS_INLINE decltype(auto) operator()(const T& t) noexcept {
    return t.size();
  }
};

/// \ingroup UtilitiesGroup
/// \brief Callable struct for the subscript operator. Returns `t[i]`
struct GetContainerElement {
  template <typename T>
  SPECTRE_ALWAYS_INLINE decltype(auto) operator()(T& t,
                                                  const size_t i) noexcept {
    return t[i];
  }
};

namespace ContainerHelpersDetail {
// implementation struct for get_element and get_size
template <bool IsFundamentalOrComplexOfFundamental>
struct GetImpls;

template <>
struct GetImpls<true> {
  template <typename T, typename SubscriptFunction>
  static SPECTRE_ALWAYS_INLINE decltype(auto) get_element(
      T& t, const size_t /*i*/, const SubscriptFunction /*at*/) noexcept {
    return t;
  }

  template <typename T, typename SizeFunction>
  static SPECTRE_ALWAYS_INLINE size_t
  get_size(const T& /*t*/, const SizeFunction /*size*/) noexcept {
    return 1;
  }
};

template <>
struct GetImpls<false> {
  template <typename T, typename SubscriptFunction>
  static SPECTRE_ALWAYS_INLINE decltype(auto) get_element(
      T& t, const size_t i, SubscriptFunction at) noexcept {
    return at(t, i);
  }

  template <typename T, typename SizeFunction>
  static SPECTRE_ALWAYS_INLINE decltype(auto) get_size(
      const T& t, SizeFunction size) noexcept {
    return size(t);
  }
};
}  // namespace ContainerHelpersDetail

template <typename T, typename SubscriptFunction = GetContainerElement>
SPECTRE_ALWAYS_INLINE decltype(auto) get_element(
    T& t, const size_t i, SubscriptFunction at = At{}) noexcept {
  return ContainerHelpersDetail::GetImpls<(
      tt::is_complex_of_fundamental_v<std::remove_cv_t<T>> or
      cpp17::is_fundamental_v<std::remove_cv_t<T>>)>::get_element(t, i,
                                                                  std::move(
                                                                      at));
}

template <typename T, typename SizeFunction = GetContainerSize>
SPECTRE_ALWAYS_INLINE decltype(auto) get_size(
    const T& t, SizeFunction size = SizeFunction{}) noexcept {
  return ContainerHelpersDetail::GetImpls<(
      tt::is_complex_of_fundamental_v<T> or
      cpp17::is_fundamental_v<T>)>::get_size(t, std::move(size));
}
```

- Fewest class creations possible
- Avoids SFINAE wherever possible
- No overly generic names in the global namespace
- Does not require users writing the structs to handle the fundamental case
- Maximally `const`
- methods are `static` on the `Impl` to avoid constructing them

moxcodes replied:
Okay, this version I'm happy with :+1: .

nilsdeppe commented on src/Utilities/ContainerHelpers.hpp:
```diff
@@ -3,43 +3,138 @@
-    -> decltype(t[i]) {
-  return t[i];
+template <typename T, typename SubscriptFunction = At>
+SPECTRE_ALWAYS_INLINE decltype(auto) get_element(
```
I have the same comments about the `get_element` that I do about `get_size`.

nilsdeppe commented on tests/Unit/Utilities/Test_ContainerHelpers.cpp:
```diff
@@ -3,34 +3,87 @@
+  double fundamental = 10.0;
+  std::complex<double> complex_of_fundamental = std::complex<double>(1.0, 5.0);
+  DataVector spectre_vector(10, 2.0);
+  CHECK(get_size(constant_stl_vector) == 10);
```
The `get_size` checks are intentionally `REQUIRE` because if the size is wrong you won't be able to sensibly check the `get_element` function.

moxcodes replied:
Wouldn't it be preferable to continue the checks in case the `get_size` merely erroneously returns 1 but the container is properly initialized with several elements?

nilsdeppe replied:
If you do out-of-bounds accesses that's undefined behavior and so literally the rest of the test is no longer a reasonable program and nothing that happens there should be considered, which is why I think it should be a hard error :)

nilsdeppe commented on src/Utilities/ContainerHelpers.hpp:
```diff
@@ -3,43 +3,138 @@
-SPECTRE_ALWAYS_INLINE size_t get_size(const T& /*t*/) noexcept {
-  return 1;
+template <typename T, typename SizeFunction = Size>
+SPECTRE_ALWAYS_INLINE size_t get_size(T& t,
```
I intentionally used `decltype(auto)` because some containers may return `ptrdiff_t` (which is weird, but :man_shrugging: )

nilsdeppe commented on tests/Unit/TestHelpers.hpp:
```diff
@@ -392,3 +397,121 @@ struct check_variables_approx<Variables<TagList>> {
+template <typename ElementwiseFunction, typename AtFunction = At,
+          typename SizeFunction = Size, typename... Arguments>
+void test_element_wise_function(ElementwiseFunction elementwise_function,
+                                std::tuple<Arguments...> arguments,
```
Is it not possible to do `const std::tuple<...>&` so that it is not copied and is const? Currently there is a lot of copying of the `tuple` going on, which is not clear to me will be cheap since I expect it to hold `DataVector`-like objects.

moxcodes replied:
I agree it should be a reference, but const is not available because it needs to work for assignment operators.

nilsdeppe replied:
Okay, so either a comment explaining that or a `not_null` would be fine with me. I have no preference :)

nilsdeppe commented on tests/Unit/TestHelpers.hpp:
```diff
@@ -392,3 +397,121 @@ struct check_variables_approx<Variables<TagList>> {
+    CHECK(custom_approx(get_element(result, i, at)) ==
+          elementwise_function(std::get<Is>(element_of_arguments)...));
+    // ensure that the final state of the arguments matches
+    expand_pack(
```
Do you think it might be clearer if the pack is expanded in order? `expand_pack` does not guarantee left-to-right evaluation of the pack. If you think it would be better to expand the pack in order then you can use the `EXPAND_PACK_LEFT_TO_RIGHT` macro. I'll leave it up to you to decide what you think is best for this case :)

nilsdeppe commented on tests/Unit/TestHelpers.hpp:
```diff
@@ -392,3 +397,121 @@ struct check_variables_approx<Variables<TagList>> {
+  auto result = elementwise_function(std::get<Is>(arguments)...);
+  for (size_t i = 0; i < size_value; ++i) {
+    // the arguments which modify their arguments must be passed lvalues
+    auto element_of_arguments = std::make_tuple(
```
`const`?

moxcodes replied:
some of these can't be const because they have to support modifying operators (like `+=`).

nilsdeppe replied:
Ah okay, thanks

nilsdeppe commented on tests/Unit/TestHelpers.hpp:
```diff
@@ -392,3 +397,121 @@ struct check_variables_approx<Variables<TagList>> {
+  });
+  // Some operators might modify the values of the arguments. We take care to
+  // verify that the modifications (or lack thereof) are also as expected.
+  auto original_arguments = std::make_tuple(std::get<Is>(arguments)...);
```
Might it be better to modify `original_arguments` so that the `std::tuple` can be `const &`? That'll reduce the number of copies going on.

moxcodes replied:
inplace operators also screw up having the originals references to the passed ones (i.e. they'd get modified in ways difficult to carefully control for).

nilsdeppe commented on tests/Unit/TestHelpers.hpp:
```diff
@@ -392,3 +397,121 @@ struct check_variables_approx<Variables<TagList>> {
+  // Some operators might modify the values of the arguments. We take care to
+  // verify that the modifications (or lack thereof) are also as expected.
+  auto original_arguments = std::make_tuple(std::get<Is>(arguments)...);
+  auto result = elementwise_function(std::get<Is>(arguments)...);
```
`const`?

nilsdeppe commented on tests/Unit/TestHelpers.hpp:
```diff
@@ -392,3 +397,121 @@ struct check_variables_approx<Variables<TagList>> {
+                                     Approx custom_approx) noexcept {
+  size_t size_value = std::max({get_size(std::get<Is>(arguments), size)...});
+  tuple_fold(arguments, [&size, &size_value ](const auto x) noexcept {
+    ASSERT(get_size(x, size) == size_value or get_size(x, size) == 1,
```
Why not have this check be active in Release mode as well by making it an:
```cpp
if (...) {
  ERROR(...);
}
```
I think it's important to do that in this case because the rest of the test is nonsensical if this check fails :)

I'd also be fine with using Catch's `REQUIRE` macro + `INFO` if you want a message :)

nilsdeppe commented on tests/Unit/TestHelpers.hpp:
```diff
@@ -392,3 +397,121 @@ struct check_variables_approx<Variables<TagList>> {
+                                     IndexingFunction at, SizeFunction size,
+                                     std::index_sequence<Is...> /*meta*/,
+                                     Approx custom_approx) noexcept {
+  size_t size_value = std::max({get_size(std::get<Is>(arguments), size)...});
```
`const`

nilsdeppe commented on tests/Unit/TestHelpers.hpp:
```diff
@@ -392,3 +397,121 @@ struct check_variables_approx<Variables<TagList>> {
+ * \ingroup TestingFrameworkGroup
+ * \brief Utility function for verifying the action of an element wise function
+ * on containers, or on some combination of containers and compatible
+ * non-containers (e.g. DataVectors with doubles).
```
Please put ticks: "`DataVector`s with `double`s" :)

nilsdeppe commented on tests/Unit/TestHelpers.hpp:
```diff
@@ -392,3 +397,121 @@ struct check_variables_approx<Variables<TagList>> {
+ * on containers, or on some combination of containers and compatible
+ * non-containers (e.g. DataVectors with doubles).
+ *
+ * \details The ability to specify custom functions for `at` and `size_of` is
```
`size_of` -> `size`?

nilsdeppe commented on tests/Unit/DataStructures/Test_VectorImplTestHelper.cpp:
```diff
@@ -4,10 +4,8 @@
 #include "tests/Unit/TestingFramework.hpp"
 
 #include <array>
-#include <complex>
```
I don't think this removes unnecessary `typename`s anymore, so could you update the commit message please? :)

nilsdeppe commented on tests/Unit/DataStructures/VectorImplTestHelper.hpp:
```diff
@@ -366,5 +376,338 @@ void vector_test_math_after_move(
   }
 }
+
+using Bound = std::array<double, 2>;
```
Is this actually used? It seems a bit out of place

moxcodes replied:
Used in `Test_DataVector.cpp`. Allows for the distribution bounds to be declared in a more expressive way. Do you have a suggestion for where in this file it should be instead?

nilsdeppe replied:
I would put it in the file it's used. As is it's sort of this undocumented feature that appears out of nowhere :)

moxcodes replied:
I prefer it in the test helper generally, as all of the various vector tests (e.g. `Test_ComplexDataVector` etc) will also need to define distribution bounds; is it better if I just document its intent with a code comment? Ultimately, as it is only a single line, it would be fine if you were to insist on duplicating it in each test.

nilsdeppe replied:
Since the alias is in a namespace I'm fine with it being left and documented as you did :)

nilsdeppe commented on tests/Unit/DataStructures/VectorImplTestHelper.hpp:
```diff
@@ -366,5 +376,338 @@ void vector_test_math_after_move(
+using WrapperList =
+    tmpl::list<UseRefWrap::Cref, UseRefWrap::None, UseRefWrap::Ref>;
+
+using NonConstWrapperList = tmpl::list<UseRefWrap::None, UseRefWrap::Ref>;
```
It seems safer to do:
```cpp
using NonConstWrapperList = tmpl::list<UseRefWrap::None, UseRefWrap::Ref>;

using WrapperList = tmpl::push_front<NonConstWrapperList, UserRefWrap::Cref>;
```
(or `push_back` if you prefer :) )

nilsdeppe commented on tests/Unit/DataStructures/VectorImplTestHelper.hpp:
```diff
@@ -366,5 +376,338 @@ void vector_test_math_after_move(
+struct VectorOrArrayAt {
+  // array of vectors version
+  template <typename T, size_t S>
+  auto operator()(std::array<T, S>& container, size_t index) noexcept {
```
You want these to have `decltype(auto)` or `T&` as the return type, with the appropriate `const` where necessary if you go for the `T&` approach :)

nilsdeppe commented on tests/Unit/DataStructures/VectorImplTestHelper.hpp:
```diff
@@ -366,5 +376,338 @@ void vector_test_math_after_move(
+  const DataVector used_for_size{size_distribution(generator)};
+  // using each distribution, generate a value for the appropriate operand type
+  // and put it in a tuple.
+  auto operand_values = std::make_tuple(make_with_random_values<Operands>(
```
Can these variables be `const`?

moxcodes replied:
Afraid not these ones. Trial and error indicates that these propagate to consts inside the wrappers, screwing things up for modifying operators like `+=`.

nilsdeppe replied:
Okay :)

nilsdeppe commented on tests/Unit/DataStructures/VectorImplTestHelper.hpp:
```diff
@@ -366,5 +376,338 @@ void vector_test_math_after_move(
+      function, wrapped_operands, VectorOrArrayAt{}, VectorOrArraySize{});
+}
+
+// enum for clarity in integral_constant below for choosing between execution
```
Is this comment outdated referring to an "enum"?

nilsdeppe commented on tests/Unit/DataStructures/VectorImplTestHelper.hpp:
```diff
@@ -366,5 +376,338 @@ void vector_test_math_after_move(
+ *  - `TestKind::Inplace`: executed on all combinations of arguments after the
+ *    first, so first is always the 'left hand side' of the operator. In this
+ *    case, at least two `VectorTypes` must be specified, where the first is
+ * used only for the left-hand side.
```
Indentation on this line for consistency :)

nilsdeppe commented on tests/Unit/DataStructures/VectorImplTestHelper.hpp:
```diff
@@ -366,5 +376,338 @@ void vector_test_math_after_move(
+ *  - `TestKind::Normal`: executed on all combinations of arguments and value
+ *    types
+ *  - `TestKind::Strict`: executed on all combinations of arguments, for only
+ * the vector types
```
Indentation on this line for consistency :)

nilsdeppe commented on tests/Unit/DataStructures/VectorImplTestHelper.hpp:
```diff
@@ -366,5 +376,338 @@ void vector_test_math_after_move(
+ *    case, at least two `VectorTypes` must be specified, where the first is
+ * used only for the left-hand side.
+ *
+ * \tparam VectorType0 The first vector type for which conmbinations are
```
"conmbinations" -> "combinations" (you have an extra "n" before the "m"

nilsdeppe commented on tests/Unit/DataStructures/VectorImplTestHelper.hpp:
```diff
@@ -366,5 +376,338 @@ void vector_test_math_after_move(
+        Test == TestKind::Strict,
+        // if strict, the operand choices are just the vector types passed in
+        tmpl::remove_duplicates<tmpl::list<VectorType0, VectorTypes...>>,
+        // else, the oberand choices include both the vectors and their element
```
"oberand" -> "operand"

nilsdeppe commented on tests/Unit/DataStructures/VectorImplTestHelper.hpp:
```diff
@@ -366,5 +376,338 @@ void vector_test_math_after_move(
+      const Function& function, const std::tuple<DistBounds...>& bounds,
+      const std::tuple<Wraps...>& wraps,
+      const std::tuple<Operands...>&& operands,
+      TestFunctionsWithVectorArgumentsChoices::Done /*done*/) noexcept {
```
The `/*done*/` etc. commented out arg names we've typically been named `/*meta*/` to indicate it's used for metaprogramming only :)

nilsdeppe reviewed (changes requested):
Overall this looks good, no major change requests :) Would you mind adding fixup terms? Thanks :)

moxcodes commented:
Fixups posted (aside from the comments under discussion). Thanks for the review!

nilsdeppe commented on src/Utilities/ContainerHelpers.hpp:
```diff
@@ -61,7 +61,7 @@ struct GetSizeImpl;
 struct GetSizeImpl<T, true> {
   template <typename SizeFunction>
-  decltype(auto) operator()(T& /*t*/, SizeFunction /*size*/) noexcept {
+  size_t operator()(T& /*t*/, SizeFunction /*size*/) noexcept {
```
I think this will go away anyway :)

moxcodes replied:
not a git mishap, more a compiler mishap. Without the `size_t` return here, I get a bunch of warnings from all over the code base as this ends up returning an int

moxcodes replied:
oh... I see.. wrong commit >.<

nilsdeppe reviewed (commented):
One git mishap that I spotted, and resolving our discussion on the last couple things. I'm happy for you to squash :)

moxcodes commented:
I've squashed the "test elementwise..." fixup and updated the other fixups with the `REQUIRE` and a comment for the `Bound` alias. I'll wait until discussions are complete on the last couple of items before squashing the final fixups :).
Cheers!

kidder commented on src/Utilities/ContainerHelpers.hpp:
```diff
@@ -3,43 +3,138 @@
 #include "Utilities/Requires.hpp"
 
+/// \ingroup UtilitiesGroup
+/// \brief Callable struct which fretrieves the `t.size()` for operand `t`. This
```
retrieves

kidder commented on src/Utilities/ContainerHelpers.hpp:
```diff
@@ -3,43 +3,138 @@
+              tt::is_complex_of_fundamental_v<std::remove_cv_t<T>> or
+              cpp17::is_fundamental_v<std::remove_cv_t<T>>)>
+struct GetElementImpl;
+template <typename T>
```
blank lines between structs

kidder commented on src/Utilities/ContainerHelpers.hpp:
```diff
@@ -3,43 +3,138 @@
+ * `gsl::at`.
+ * \note `std::complex` are regarded as non-indexable (despite a predictable
+ * memory layout), so this function acts as the identity on `std::complex` of
+ * fundamental types)
```
unbalanced `)`

kidder commented on src/Utilities/ContainerHelpers.hpp:
```diff
@@ -3,43 +3,138 @@
+ * `std::size()` as of C++17.
+ * \note `std::complex` are regarded as non-indexable (despite a predictable
+ * memory layout), so this function will return 1 for a `std::complex` of a
+ * fundamental type)
```
unbalanced `)`

kidder commented on tests/Unit/TestHelpers.hpp:
```diff
@@ -393,3 +397,128 @@ struct check_variables_approx<Variables<TagList>> {
+
+/*!
+ * \ingroup TestingFrameworkGroup
+ * \brief A test utility for verifying that an element wise function, `function`
```
`elementwise` or `element-wise` everywhere

kidder commented:
I am happy for you to squash if @nilsdeppe is

nilsdeppe commented:
I'm happy for it to be squashed. I'll get back to you on the last two outstanding comments in a bit :)

moxcodes commented:
fixups squashed in.
Cheers!

nilsdeppe commented on tests/Unit/TestHelpers.hpp:
```diff
@@ -393,3 +397,128 @@ struct check_variables_approx<Variables<TagList>> {
+template <typename Function, typename IndexingFunction, typename SizeFunction,
+          typename... Arguments, size_t... Is>
+void test_element_wise_function_impl(Function elementwise_function,
+                                     std::tuple<Arguments...>& arguments,
```
since the `arguments` are being modified they should be taken by `gsl::not_null`

nilsdeppe reviewed (commented):
I'm fine with you squashing the fixups

moxcodes commented:
The last couple of fixes are squashed in. The only changes are the requested alteration to `ContainerHelpers.hpp` and the `not_null` on the element-wise check utility, and propagating changes to the rest of the code so everything still works.
Cheers!

nilsdeppe commented on src/Utilities/ContainerHelpers.hpp:
```diff
@@ -3,43 +3,134 @@
+template <typename T, typename SizeFunction = GetContainerSize>
+SPECTRE_ALWAYS_INLINE decltype(auto) get_size(
+    T& t, SizeFunction size = GetContainerSize{}) noexcept {
+  return ContainerHelpersDetail::GetImpls <
```
The formatting here is normal if you put the `tt:: or cpp17::` in parens :)

nilsdeppe commented on src/Utilities/ContainerHelpers.hpp:
```diff
@@ -3,43 +3,134 @@
+template <typename T, typename SubscriptFunction = GetContainerElement>
+SPECTRE_ALWAYS_INLINE decltype(auto) get_element(
+    T& t, size_t i, SubscriptFunction at = GetContainerElement{}) noexcept {
+  return ContainerHelpersDetail::GetImpls <
```
Same formatting comment as for `get_size`

nilsdeppe commented on src/Utilities/ContainerHelpers.hpp:
```diff
@@ -3,43 +3,134 @@
-  return t;
+template <typename T, typename SubscriptFunction = GetContainerElement>
+SPECTRE_ALWAYS_INLINE decltype(auto) get_element(
+    T& t, size_t i, SubscriptFunction at = GetContainerElement{}) noexcept {
```
`i` can be `const`

nilsdeppe commented on src/Utilities/ContainerHelpers.hpp:
```diff
@@ -3,43 +3,134 @@
-  return 1;
+template <typename T, typename SizeFunction = GetContainerSize>
+SPECTRE_ALWAYS_INLINE decltype(auto) get_size(
+    T& t, SizeFunction size = GetContainerSize{}) noexcept {
```
`T&` can be `const T&` (getting the size shouldn't alter the container)

nilsdeppe commented on src/Utilities/ContainerHelpers.hpp:
```diff
@@ -3,43 +3,134 @@
+/// \brief Callable struct for the subscript operator. Returns `t[i]`
+struct GetContainerElement {
+  template <typename T>
+  SPECTRE_ALWAYS_INLINE decltype(auto) operator()(T& t, size_t i) noexcept {
```
`i` can be `const`

nilsdeppe commented on tests/Unit/TestHelpers.hpp:
```diff
@@ -392,3 +397,126 @@ struct check_variables_approx<Variables<TagList>> {
+          typename SizeFunction = GetContainerSize, typename... Arguments>
+void test_element_wise_function(
+    ElementwiseFunction elementwise_function,
+    gsl::not_null<std::tuple<Arguments...>*> arguments,
```
`const gsl::not_null` (the pointer shouldn't change, just the data pointed to)

nilsdeppe commented on tests/Unit/TestHelpers.hpp:
```diff
@@ -392,3 +397,126 @@ struct check_variables_approx<Variables<TagList>> {
+          typename... Arguments, size_t... Is>
+void test_element_wise_function_impl(
+    Function elementwise_function,
+    gsl::not_null<std::tuple<Arguments...>*> arguments, IndexingFunction at,
```
`const gsl::not_null`

nilsdeppe reviewed (commented):
A few very minor comments requesting more `const` and dealing with clang-format making weird formatting. Please squash again :)

@kidder could you please give feedback?

moxcodes commented:
squashed in again.
Cheers!

wthrowe commented on src/Utilities/ContainerHelpers.hpp:
```diff
@@ -3,43 +3,136 @@
+  }
+};
+
+namespace ContainerHelpersDetail {
```
`_detail` (hides from documentation)

wthrowe commented on src/Utilities/ContainerHelpers.hpp:
````diff
@@ -3,43 +3,136 @@
+ * instance, for a type which is a `std::array` of a `std::vector`, the indexing
+ * function could be the below callable struct:
+ * ```
+ * struct ArrayOfVectorIndexFunctor {
````
Make this a snippet.  (And use in the test enough to catch the current problem with the code.)

wthrowe commented on src/Utilities/ContainerHelpers.hpp:
````diff
@@ -3,43 +3,136 @@
+ * instance, for a type which is a `std::array` of a `std::array`, the size
+ * function could be the below callable struct:
+ * ```
+ * struct ArrayOfArraysSizeFunctor {
````
Snippet again.  (And there is also a problem here, but the test probably won't catch it in this case.)

wthrowe commented on tests/Unit/TestHelpers.hpp:
```diff
@@ -392,3 +397,126 @@ struct check_variables_approx<Variables<TagList>> {
+                               size_of_operator);                \
+  } while (false)
+
+namespace TestHelpersDetail {
```
`_detail`

wthrowe replied:
ping

wthrowe commented on tests/Unit/TestHelpers.hpp:
```diff
@@ -392,3 +397,126 @@ struct check_variables_approx<Variables<TagList>> {
+          typename AtFunction = GetContainerElement,
+          typename SizeFunction = GetContainerSize, typename... Arguments>
+void test_element_wise_function(
+    ElementwiseFunction elementwise_function,
```
You are frequently inconsistent about whether "element-wise" is one word or two.  (`test_element_wise_function(elementwise_function)` stood out particularly here.)  I'm not going to block a merge on that, but you can adjust it if you want to.

wthrowe commented on tests/Unit/DataStructures/VectorImplTestHelper.hpp:
```diff
@@ -366,5 +376,343 @@ void vector_test_math_after_move(
+
+/*!
+ * \ingroup TestingFrameworkGroup
+ * \brief General entry function for testing arbitrary math functions
```
This isn't documenting a function, although it would seem more natural for this thing to be a function and it isn't obvious to me why it isn't.

wthrowe commented on tests/Unit/DataStructures/Test_DataVector.cpp:
```diff
@@ -53,353 +50,97 @@
+
+  const auto strict_binary_ops = std::make_tuple(
+      std::make_tuple(funcl::Atan2<>{}, std::make_tuple(generic, positive)),
+      std::make_tuple(funcl::Hypot<>{}, std::make_tuple(generic, positive)));
```
Why positive for the second argument? (atan2 I guess is to avoid the origin.)

moxcodes commented:
Fixups posted. I believe I've addressed all of the remaining comments.
Cheers!

wthrowe commented on src/Utilities/ContainerHelpers.hpp:
````diff
@@ -75,17 +75,10 @@ struct GetImpls<false> {
- * ```
+ * \snippet Test_ContainerHelpers.cpp get_element_example_indexing_callable
+ *
  * which would index the data structure in a manner in which the array index
````
There are now two array indices, so "the array index" is ambiguous.

moxcodes commented:
Just pushed a couple more fixups for the new comments on the items I missed.
Cheers!

wthrowe commented:
I'm happy with this.  Squash.

moxcodes commented:
fixups squashed in.
Cheers!
